File: apps/grn/src/types/listing.ts
export interface CatalogCrop {
  id: string;
  slug: string;
  commonName: string;
  scientificName: string | null;
  category: string | null;
  description: string | null;
  /** Garden pyramid layer 1=Foundation..5=Joy; null when not part of the food pyramid. */
  pyramidTier?: number | null;
}

export interface CatalogVariety {
  id: string;
  cropId: string;
  slug: string;
  name: string;
  description: string | null;
}

export interface GrowerCropItem {
  id: string;
  userId: string;
  canonicalId: string | null;
  cropName: string;
  varietyId: string | null;
  status: string;
  visibility: string;
  surplusEnabled: boolean;
  nickname: string | null;
  defaultUnit: string | null;
  notes: string | null;
  bedId: string | null;
  bedName: string | null;
  plantingDate: string | null;
  expectedHarvestDate: string | null;
  plantCount: number | null;
  spacingInches: number | null;
  /** Garden pyramid layer resolved from the linked catalog crop; null when unlinked or not in the pyramid. */
  pyramidTier?: number | null;
  createdAt: string;
  updatedAt: string;
}

export type SunExposure =
  | 'full_sun'
  | 'partial_sun'
  | 'partial_shade'
  | 'full_shade'
  | 'mixed';

export type BedType = 'in_ground' | 'raised' | 'mound';
export type BedShape = 'rect' | 'circle' | 'polygon';

export interface BedPolygonPoint {
  x: number;
  y: number;
}

export interface GardenBed {
  id: string;
  userId: string;
  name: string;
  description: string | null;
  sunExposure: SunExposure | null;
  soilType: string | null;
  lengthInches: number | null;
  widthInches: number | null;
  locationNotes: string | null;
  sortOrder: number;
  bedType: BedType;
  shape: BedShape;
  positionX: number | null;
  positionY: number | null;
  rotationDeg: number;
  points: BedPolygonPoint[] | null;
  color: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface GardenCanvas {
  id: string;
  userId: string;
  widthInches: number;
  heightInches: number;
  backgroundImageKey: string | null;
  backgroundImageUrl: string | null;
  backgroundOpacity: number;
  northOffsetDeg: number;
  createdAt: string;
  updatedAt: string;
}

export type AnnotationShape = 'rect' | 'circle' | 'polygon' | 'line';

export interface GardenAnnotation {
  id: string;
  userId: string;
  label: string;
  icon: string | null;
  shape: AnnotationShape;
  positionX: number | null;
  positionY: number | null;
  lengthInches: number | null;
  widthInches: number | null;
  rotationDeg: number;
  points: BedPolygonPoint[] | null;
  color: string | null;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface Listing {
  id: string;
  userId: string;
  growerCropId: string | null;
  cropId: string;
  varietyId: string | null;
  title: string;
  unit: string;
  quantityTotal: string;
  quantityRemaining: string;
  availableStart: string;
  availableEnd: string;
  status: string;
  pickupLocationText: string | null;
  pickupAddress: string | null;
  pickupDisclosurePolicy: string;
  pickupNotes: string | null;
  contactPref: string;
  geoKey: string | null;
  lat: number;
  lng: number;
  createdAt: string;
}

export interface ListMyListingsResponse {
  items: Listing[];
  limit: number;
  offset: number;
  hasMore: boolean;
  nextOffset: number | null;
}

export interface DiscoverListingsResponse {
  items: Listing[];
  limit: number;
  offset: number;
  hasMore: boolean;
  nextOffset: number | null;
}

export interface UpsertListingRequest {
  title: string;
  cropId?: string;  // For catalog crops
  growerCropId?: string;  // For user-defined crops
  varietyId?: string;
  quantityTotal: number;
  unit: string;
  availableStart: string;
  availableEnd: string;
  pickupLocationText?: string;
  pickupAddress?: string;
  pickupDisclosurePolicy?: 'immediate' | 'after_confirmed' | 'after_accepted';
  pickupNotes?: string;
  contactPref?: 'app_message' | 'phone' | 'knock';
  lat: number;
  lng: number;
  status?: 'active' | 'pending' | 'claimed' | 'expired' | 'completed';
}

File: services/grn-api/src/api/models/listing.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListingItem {
    pub id: String,
    pub user_id: String,
    pub grower_crop_id: Option<String>,
    pub crop_id: Option<String>,
    pub variety_id: Option<String>,
    pub title: Option<String>,
    pub unit: Option<String>,
    pub quantity_total: Option<String>,
    pub quantity_remaining: Option<String>,
    pub available_start: Option<String>,
    pub available_end: Option<String>,
    pub status: String,
    pub pickup_location_text: Option<String>,
    pub pickup_address: Option<String>,
    pub effective_pickup_address: Option<String>,
    pub pickup_disclosure_policy: String,
    pub pickup_notes: Option<String>,
    pub contact_pref: String,
    pub geo_key: Option<String>,
    pub lat: Option<f64>,
    pub lng: Option<f64>,
    pub created_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListMyListingsResponse {
    pub items: Vec<ListingItem>,
    pub limit: i64,
    pub offset: i64,
    pub has_more: bool,
    pub next_offset: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverListingsResponse {
    pub items: Vec<ListingItem>,
    pub limit: i64,
    pub offset: i64,
    pub has_more: bool,
    pub next_offset: Option<i64>,
}

#[cfg(test)]
mod tests {
    use super::ListingItem;

    #[test]
    fn custom_crop_listing_serializes_without_a_catalog_crop() {
        let listing = ListingItem {
            id: "listing-id".to_string(),
            user_id: "user-id".to_string(),
            grower_crop_id: Some("grower-crop-id".to_string()),
            crop_id: None,
            variety_id: None,
            title: Some("Purple yard beans".to_string()),
            unit: Some("bunch".to_string()),
            quantity_total: Some("3".to_string()),
            quantity_remaining: Some("3".to_string()),
            available_start: Some("2026-07-15T12:00:00Z".to_string()),
            available_end: Some("2026-07-17T12:00:00Z".to_string()),
            status: "active".to_string(),
            pickup_location_text: Some("Front porch".to_string()),
            pickup_address: None,
            effective_pickup_address: Some("1100 Congress Ave, Austin, TX 78701".to_string()),
            pickup_disclosure_policy: "after_confirmed".to_string(),
            pickup_notes: None,
            contact_pref: "app_message".to_string(),
            geo_key: Some("9v6kpv".to_string()),
            lat: Some(30.2747),
            lng: Some(-97.7404),
            created_at: "2026-07-15T12:00:00Z".to_string(),
        };

        let serialized = serde_json::json!(listing);

        assert_eq!(serialized["growerCropId"], "grower-crop-id");
        assert!(serialized["cropId"].is_null());
    }
}
